File: services/product-service/internal/domain/product.go
package domain

import (
	"context"
	"time"
)

type Product struct {
	ID          int64
	Name        string
	Slug        string
	Description string
	Metadata    map[string]interface{} // JSON object
	Features    map[string]interface{} // JSON object
	IsPopular   bool
	IsDefault   bool
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

type ProductRepository interface {
	GetByID(ctx context.Context, id int64) (*Product, error)
	GetBySlug(ctx context.Context, slug string) (*Product, error)
	Create(ctx context.Context, product *Product) error
	Update(ctx context.Context, product *Product) error
	Delete(ctx context.Context, id int64) error
	GetAll(ctx context.Context, page, perPage int) ([]*Product, int, error)
}

type ProductService interface {
	GetByID(ctx context.Context, id int64) (*Product, error)
	GetBySlug(ctx context.Context, slug string) (*Product, error)
	Create(ctx context.Context, product *Product) error
	Update(ctx context.Context, product *Product) error
	Delete(ctx context.Context, id int64) error
	GetAll(ctx context.Context, page, perPage int) ([]*Product, int, error)
}

File: web/src/types/product.ts
export interface Product {
  id: string
  name: string
  slug: string
  description?: string
  metadata?: string
  features?: string
  isPopular: boolean
  isDefault: boolean
  createdAt: number
  updatedAt: number
}

export interface ProductsResponse {
  products: {
    success: boolean
    message: string
    data: {
      products: Product[]
      total: number
      page: number
      perPage: number
    }
  }
}

export interface ProductResponse {
  product: {
    success: boolean
    message: string
    data: Product
  }
}

export interface CreateProductInput {
  name: string
  slug: string
  description?: string
  metadata?: string
  features?: string
  isPopular?: boolean
  isDefault?: boolean
}

export interface UpdateProductInput {
  id: string
  name?: string
  slug?: string
  description?: string
  metadata?: string
  features?: string
  isPopular?: boolean
  isDefault?: boolean
}

export interface CreateProductResponse {
  createProduct: {
    success: boolean
    message: string
    data: Product
  }
}

export interface UpdateProductResponse {
  updateProduct: {
    success: boolean
    message: string
    data: Product
  }
}

export interface DeleteProductResponse {
  deleteProduct: {
    success: boolean
    message: string
  }
}
